tests/node_runtime.rs

```diff
@@ -1,78 +1,97 @@
-use frame_system::offchain::{SendSignedTransaction, Signer};
-use futures::compat::Future01CompatExt;
-use pallet_balances::Call as BalancesCall;
-use node_runtime::{Runtime, RuntimeApi};
-use sp_core::crypto::Pair;
-use sp_keyring::Sr25519Keyring;
-use sp_runtime::{traits::IdentifyAccount, MultiSigner, MultiSignature};
-use substrate_test_runner::{
-	prelude::*, rpc, test, node::{InternalNode, TestRuntimeRequirements},
-};
-use sc_service::{TFullClient, new_full_parts, TFullBackend, TaskManager, Configuration};
-use parity_scale_codec::alloc::sync::Arc;
-use sc_keystore::KeyStorePtr;
+
+//! End to end runtime tests
+
+use substrate_test_runner::{Node, ChainInfo, SignatureVerificationOverride};
+use grandpa::GrandpaBlockImport;
+use sc_service::{TFullBackend, TFullClient, Configuration, TaskManager, new_full_parts};
+use node_runtime::{SignedExtra, Runtime, Call, Event};
+use sp_runtime::generic::Era;
+use std::{sync::Arc, str::FromStr};
 use sp_inherents::InherentDataProviders;
 use sc_consensus_babe::BabeBlockImport;
-use sc_finality_grandpa::GrandpaBlockImport;
-use manual_seal::consensus::{ConsensusDataProvider,	babe::BabeConsensusDataProvider};
-use sp_api::TransactionFor;
-use sp_application_crypto::sr25519;
-
-struct Node;
+use sp_keystore::SyncCryptoStorePtr;
+use sp_keyring::sr25519::Keyring::Alice;
+use node_cli::chain_spec::development_config;
+use sp_consensus_babe::AuthorityId;
+use sc_consensus_manual_seal::{ConsensusDataProvider, consensus::babe::BabeConsensusDataProvider};
+use sp_runtime::AccountId32;
 
 type BlockImport<B, BE, C, SC> = BabeBlockImport<B, C, GrandpaBlockImport<BE, B, C, SC>>;
 
-impl TestRuntimeRequirements for Node {
+sc_executor::native_executor_instance!(
+	pub Executor,
+	node_runtime::api::dispatch,
+	node_runtime::native_version,
+	SignatureVerificationOverride,
+);
+
+/// ChainInfo implementation.
+struct NodeTemplateChainInfo;
+
+impl ChainInfo for NodeTemplateChainInfo {
 	type Block = node_primitives::Block;
-	type Executor = node_executor::Executor;
-	type Runtime = Runtime;
-	type RuntimeApi = RuntimeApi;
+	type Executor = Executor;
+	type Runtime = node_runtime::Runtime;
+	type RuntimeApi = node_runtime::RuntimeApi;
 	type SelectChain = sc_consensus::LongestChain<TFullBackend<Self::Block>, Self::Block>;
 	type BlockImport = BlockImport<
 		Self::Block,
 		TFullBackend<Self::Block>,
 		TFullClient<Self::Block, Self::RuntimeApi, Self::Executor>,
 		Self::SelectChain,
 	>;
+	type SignedExtras = SignedExtra;
 
 	fn load_spec() -> Result<Box<dyn sc_service::ChainSpec>, String> {
-		Ok(Box::new(node_cli::chain_spec::development_config()))
+		Ok(Box::new(development_config()))
+	}
+
+	fn signed_extras(from: <Self::Runtime as frame_system::Config>::AccountId) -> Self::SignedExtras {
+		(
```

Comment:
I feel this is something that should be coming from the runtime, we do something very similar in `CreateSignedTransaction` trait already. In the future it would be good to DRY this somehow.

Reply:
well the problem is signed extensions can be user-defined, and so may take inputs that substrate may not account for,

Reply:
Yeah, but we can easily create a generic runtime API that looks like this:
```rust
fn create_transaction(parameters: Map<String, Value>) -> SignedExtensions {
   let tip = parameters.get("tip").parseAsNumber()?;
}
```

As I said, not a priority right now, but something to consider long-term, cause every client runs into exact same problem every time and we keep replicating the same solution of hardcoding signed extensions.